### PR TITLE
Make DTensor.local_map pytree-aware on in/out_placements

### LINKED ISSUE #182031: DTensor.local_map inconsistently handles pytree on input versus placements
The current behavior is that it will pytree flatten the arguments, but it treats placements as a flat list to compare against the flattened one. This seems like the worst of both worlds.

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx @zpcore

### PR DESCRIPTION
**Pull request by aditvenk**
### Summary
`torch.distributed.tensor.experimental.local_map` pytree-flattens its
positional arguments via `pytree.tree_flatten(args)`, but it then treats
`in_placements`, `in_grad_placements`, and `out_placements` as flat sequences
to be compared by length against the flattened args/output. This is the
"worst of both worlds" called out in the issue: callers who pass nested
structures (e.g., a list of DTensors as a single positional argument) cannot
mirror that structure in the placements; they must mentally flatten the
structure themselves.

### Root Cause
In `_local_map_wrapped` in
`torch/distributed/tensor/experimental/_func_map.py`:

1. Args were pytree-flattened to `flat_args` and the loop indexed
 `in_placements[idx]` with that flattened index.
2. The output path normalized `out_placements` only with `(... if
 isinstance(... , tuple) else (... ,))`, never recursing into nested
 structures.

So nested args produced more flat leaves than the user-natural per-arg
`in_placements` tuple, leading to errors like
`in_placements length 1 does not match the number of input args 2!`.

### Fix
Make all three placement structures pytree-aware with `is_leaf` detection of
a `PlacementType` (`None` or a non-empty `Sequence[Placement]`):

- `_is_placement_spec(x)` returns `True` for `None` or a non-empty list/tuple
 whose elements are all `Placement` instances.
- `_flatten_input_placements(placements, num_flat_args, label)` pytree-flattens
 `in_placements` / `in_grad_placements` using that `is_leaf` and verifies the
 resulting flat-leaf count matches `len(flat_args)`.
- `_flatten_output_placements(out_placements, num_flat_out)` does the same for
 `out_placements`, naturally handling both the single-leaf shorthand
 (`out_placements=[Shard(0)]`) and tuples of placement specs.

Existing flat usages remain valid because each top-level placement spec
(`None` or `[Shard(0)]`, `[Replicate()]`, etc.) is itself a leaf under the
`is_leaf` rule, so flattening is a no-op for them. The new behavior accepts
nested structures such as `in_placements=([row_wise, row_wise],)` or
`in_placements=({"a": row_wise, "b": row_wise},)`, which mirror the args
pytree. The docstring is updated to describe both forms.

The change is in pure Python; no rebuild is required.

### Repro
Run:

```bash
python repro_182031_generated.py
```

<details>
<summary>Repro Script</summary>

```python
"""
Repro for issue #182031: DTensor.local_map inconsistently handles pytree on
input versus placements.

The bug: `local_map` pytree-flattens its positional arguments but then expects
in_placements (and out_placements) to be a flat tuple matching the length of
the flattened args. So a user who passes a nested structure (e.g., a list of
DTensors) cannot use a nested placements spec that mirrors the arg structure;
they are forced to mentally flatten the structure first.

After the fix, in_placements and out_placements may either be flat (the old
behavior, preserved) OR may mirror the pytree structure of the args/output.
Both flatten to the same set of leaves, paired 1:1 with the pytree-flattened
args/output values. A leaf is a `PlacementType`: `None` or a `Sequence` of
`Placement`.
"""

import os
import sys
import traceback

import torch
import torch.distributed as dist
from torch.distributed.tensor import distribute_tensor, Replicate, Shard
from torch.distributed.tensor.experimental import local_map


def main():
 os.environ.setdefault("MASTER_ADDR", "localhost")
 os.environ.setdefault("MASTER_PORT", "29555")
 os.environ.setdefault("RANK", "0")
 os.environ.setdefault("WORLD_SIZE", "1")

 dist.init_process_group(backend="gloo")
 try:
 device_mesh = dist.device_mesh.init_device_mesh("cpu", (1,))

 row_wise = [Shard(0)]
 replicate = [Replicate()]

 X = torch.randn(8, 8)
 Y = torch.randn(8, 8)
 X_dt = distribute_tensor(X, device_mesh, row_wise)
 Y_dt = distribute_tensor(Y, device_mesh, row_wise)

 def add_pair(pair):
 # pair is [t1, t2]; return their sum as a single tensor.
 return pair[0] + pair[1]

 def split_pair(pair):
 # pair is [t1, t2]; return a list (nested output).
 return [pair[0] * 2.0, pair[1] * 3.0]

 # Test 1: flat in_placements still works (backward compat).
 flat_fn = local_map(
 add_pair,
 out_placements=row_wise,
 in_placements=(row_wise, row_wise), # flat
 device_mesh=device_mesh,
 )
 r1 = flat_fn([X_dt, Y_dt])
 print("flat in_placements works.")
 assert r1.placements == tuple(row_wise)

 # Test 2: nested in_placements mirroring the args structure.
 nested_fn = local_map(
 add_pair,
 out_placements=row_wise,
 in_placements=([row_wise, row_wise],), # mirror args=([t,t],)
 device_mesh=device_mesh,
 )
 r2 = nested_fn([X_dt, Y_dt])
 print("nested in_placements mirroring args works.")
 assert r2.placements == tuple(row_wise)

 # Test 3: nested output with nested out_placements.
 nested_out_fn = local_map(
 split_pair,
 out_placements=[row_wise, row_wise], # mirror output=[t,t]
 in_placements=([row_wise, row_wise],),
 device_mesh=device_mesh,
 )
 out3 = nested_out_fn([X_dt, Y_dt])
 assert isinstance(out3, list) and len(out3) == 2
 for d in out3:
 assert d.placements == tuple(row_wise)
 print("nested output with nested out_placements works.")

 # Test 4: flat out_placements still works (backward compat).
 flat_out_fn = local_map(
 split_pair,
 out_placements=(row_wise, row_wise), # flat
 in_placements=(row_wise, row_wise),
 device_mesh=device_mesh,
 )
 out4 = flat_out_fn([X_dt, Y_dt])
 assert isinstance(out4, list) and len(out4) == 2
 for d in out4:
 assert d.placements == tuple(row_wise)
 print("flat out_placements still works.")

 # Test 5: mismatched length still raises a helpful error.
 bad_fn = local_map(
 add_pair,
 out_placements=row_wise,
 in_placements=(row_wise,), # too few specs for [t,t]
 device_mesh=device_mesh,
 )
 try:
 bad_fn([X_dt, Y_dt])
 except AssertionError as e:
 assert "in_placements length 1" in str(e)
 print("Mismatched length correctly errors:", e)
 else:
 print("BUG: mismatched length did not error.")
 sys.exit(1)

 # Test 6: dict-shaped args with dict-shaped placements.
 def dict_fn(d):
 return d["a"] + d["b"]

 dict_local = local_map(
 dict_fn,
 out_placements=row_wise,
 in_placements=({"a": row_wise, "b": row_wise},),
 device_mesh=device_mesh,
 )
 r6 = dict_local({"a": X_dt, "b": Y_dt})
 assert r6.placements == tuple(row_wise)
 print("dict-shaped args with dict-shaped placements works.")

 # Test 7: nested in_grad_placements mirroring a nested args tree.
 X_req = torch.randn(8, 8, requires_grad=True)
 Y_req = torch.randn(8, 8, requires_grad=True)
 Xr_dt = distribute_tensor(X_req, device_mesh, row_wise)
 Yr_dt = distribute_tensor(Y_req, device_mesh, row_wise)

 nested_grad_fn = local_map(
 add_pair,
 out_placements=row_wise,
 in_placements=([row_wise, row_wise],),
 in_grad_placements=([row_wise, row_wise],),
 device_mesh=device_mesh,
 redistribute_inputs=True,
 )
 out7 = nested_grad_fn([Xr_dt, Yr_dt])
 out7.to_local().sum().backward()
 assert Xr_dt.grad.placements == tuple(row_wise)
 assert Yr_dt.grad.placements == tuple(row_wise)
 print("nested in_grad_placements works.")

 print()
 print("All repro tests passed: local_map now handles pytree on")
 print("inputs and placements consistently.")
 sys.exit(0)
 finally:
 dist.destroy_process_group()


if __name__ == "__main__":
 try:
 main()
 except SystemExit:
 raise
 except Exception:
 traceback.print_exc()
 sys.exit(2)
```

</details>

### Testing
- `repro_182031_generated.py` exits 0 (all 7 user-facing tests pass).
- `python test/distributed/tensor/experimental/test_local_map.py` -> 8/8
 passed (6 pre-existing tests plus `test_nested_in_out_placements` and
 `test_local_map_with_nested_grad_placement`).
- `spin fixlint torch/distributed/tensor/experimental/_func_map.py
 test/distributed/tensor/experimental/test_local_map.py` -> no lint issues.

## Files Changed
- `torch/distributed/tensor/experimental/_func_map.py`
- `test/distributed/tensor/experimental/test_local_map.py`

Fixes #182031

---
*This PR was generated by [ptq](https://github.com/drisspg/pt_job_queue) with human review.*

### CODE REVIEW

**Comment by aditvenk** [comment]
@claude
